libraries/Microsoft.Bot.Builder.Azure/CosmosDbSqlStorage.cs

```diff
@@ -0,0 +1,215 @@
+    {
+        private readonly string databaseId;
+        private readonly string collectionId;
+        private readonly DocumentClient client;
+
+        private const string UserAgentSuffix = "Microsoft-BotFramework 4.0.0";
+
+        private static JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All
+        });
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CosmosDbSqlStorage"/> class,
+        /// using the provided CosmosDB credentials, DatabaseId and CollectionId.
+        /// </summary>
+        /// <param name="serviceEndpoint">The endpoint Uri for the service endpoint from the Azure Cosmos DB service.</param>
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        public CosmosDbSqlStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            this.databaseId = databaseId;
+            this.collectionId = collectionId;
+
+            var connectionPolicy = new ConnectionPolicy()
+            {
+                UserAgentSuffix = UserAgentSuffix
+            };
+
+            connectionPolicyConfigurator?.Invoke(connectionPolicy);
+
+            this.client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+            this.client.CreateDatabaseIfNotExistsAsync(new Database { Id = databaseId })
+                .Wait();
+
+            this.client.CreateDocumentCollectionIfNotExistsAsync(UriFactory.CreateDatabaseUri(databaseId), new DocumentCollection { Id = collectionId })
+                .Wait();
+        }
+
+        /// <summary>
+        /// Deletes the specified keys from storage.
+        /// </summary>
+        /// <param name="keys">List of keys to remove.</param>
+        /// <returns></returns>
+        public async Task Delete(params string[] keys)
+        {
+            var tasks = keys.Select(key =>
+                this.client.DeleteDocumentAsync(
```

Comment:
Are there any concurrency concerns around this? If not, could you add a comment to that effect in the code. 

the lifecycle here is that the "_client" will be used across any number of threads, and as such needs to be thread safe.

Reply:
Yes, the client is thread-safe. It is recommended to use a single instance per application/per CosmosDB 'server'. [Relevant link](https://azure.microsoft.com/en-us/blog/performance-tips-for-azure-documentdb-part-1-2/).